### DISCUSSION TITLE
Nuxt module should be a function

### DISCUSSION THREAD

**Discussion by telisjok** [nuxt/nuxt#16716] [Questions] [answered]
I am getting the following error 

```
 ERROR  Cannot start nuxt:  Nuxt module should be a function: @storyblok/nuxt,[object Object]                                                                                                                                                                                            18:49:39

  at installModule (/C:/JavascriptProjects/NUXT/website-ui/node_modules/@nuxt/kit/dist/index.mjs:338:11)
  at initNuxt (/C:/JavascriptProjects/NUXT/website-ui/node_modules/nuxt3/dist/index.mjs:939:11)
  at async load (/C:/JavascriptProjects/NUXT/website-ui/node_modules/nuxi/dist/chunks/dev.mjs:6717:9)
  at async Object.invoke (/C:/JavascriptProjects/NUXT/website-ui/node_modules/nuxi/dist/chunks/dev.mjs:6756:5)
  at async _main (/C:/JavascriptProjects/NUXT/website-ui/node_modules/nuxi/dist/chunks/index.mjs:384:7)
```

This is my `nuxt.config.js`

```
import { defineNuxtConfig } from 'nuxt3';

// https://v3.nuxtjs.org/docs/directory-structure/nuxt.config
export default defineNuxtConfig({
  build: {
    postcss: {
      postcssOptions: {
        plugins: {
          tailwindcss: {},
          autoprefixer: {},
        },
      },
    },
  },
  components: {
    dirs: [
      '~/components',
      '~/components/form-elements',
      '~/components/layout'
    ]
  },
  css: ['~/assets/css/index.scss',],
  modules: [
    ["@storyblok/nuxt", { accessToken: "YOUR_ACCESS_TOKEN" }]
  ],
  vue: {
    config: {
      devtools: true
    }
  }
});
```

I don't see anything being wrong here so i am not sure where this could be coming from.

package.json
```
{
  "private": true,
  "scripts": {
    "dev": "nuxi dev",
    "build": "nuxi build",
    "start": "node .output/server/index.mjs"
  },
  "devDependencies": {
    "@storyblok/nuxt": "^3.0.0-beta.1",
    "autoprefixer": "^10.4.0",
    "axios": "^0.24.0",
    "nuxt3": "latest",
    "postcss": "^8.4.5",
    "sass": "^1.45.0",
    "tailwindcss": "^3.0.6"
  }
}
```

The `.nuxt/` folder is empty and I am getting the following error in the tsconfig.json file
```
JSON schema for the TypeScript compiler's configuration file

Cannot read file 'c:/JavascriptProjects/NUXT/taus-website-ui/.nuxt/tsconfig.json'.ts
```

**Comment by pi0** [comment]
Can you please try recreating lockfile using `npx nuxi upgrade --force`

**Reply by telisjok** [reply]
Didn't fix anything. It just broke the application even more. I made a fresh git clone from the Git repo and everything is ok now.
I'm not sure where things got messed up.

**Reply by jseguraweb** [reply]
Did you find a proper solution for this error? I'm getting the same one and I can't find any explanation. As soon as I run `npm run dev` I get:

```
> nuxt-storyblok-bootstrap@1.0.0 dev
> nuxt

 FATAL  Module should export a function: @storyblok/nuxt                                                                                                                                

  at ModuleContainer.addModule (node_modules/@nuxt/core/dist/core.js:217:13)
  at node_modules/@nuxt/utils/dist/utils.js:639:43
  at async ModuleContainer.ready (node_modules/@nuxt/core/dist/core.js:51:7)
  at async Nuxt._init (node_modules/@nuxt/core/dist/core.js:478:5)

 ✖ Nuxt Fatal Error  TypeError: Module should export a function: @storyblok/nuxt

**Reply by jseguraweb** [reply]
I took a look again in the docs and something changed: the storyblok module must be imported in the modules like this `@storyblok/nuxt/module` and not like this anymore `@storyblok/nuxt`

**Comment by TrungRueta** [comment]
I can confirm issue raise after upgrade to `v3.0.0-27335068.804b323`. 
With me i has onlye 1 module `nuxt-windicss` which working on previous version. after upgrade it say 

```
 ERROR  Cannot start nuxt:  Nuxt module should be a function: nuxt-windicss,[object Object]                                                  21:34:21

  at installModule (node_modules/@nuxt/kit/dist/index.mjs:338:11)
  at initNuxt (node_modules/nuxt3/dist/index.mjs:940:11)
  at async load (node_modules/nuxi/dist/chunks/dev.mjs:6717:9)
  at async Object.invoke (node_modules/nuxi/dist/chunks/dev.mjs:6756:5)
  at async _main (node_modules/nuxi/dist/chunks/index.mjs:384:7)
```

Remove nuxt-windicss module bring nuxt working again. remove node_modules, package-lock, upgrade force etc not help.

downgrade to v3.0.0-27333388.c824905 working again.

**Reply by pi0** [reply]
Windicss is pending for upgrade: https://github.com/windicss/nuxt-windicss/pull/148

**Comment by pi0** [comment, marked as answer]
Regression fixed in the latest version. Sorry for the inconvenience.

**Reply by TheAlexLichter** [reply]
And also now fixed via https://github.com/nuxt/nuxt/pull/24957

**Comment by edwardcastle** [comment]
Greetings, I'm having the same issue with Nuxt v3.0.0-rc.3,
package.json:
{
  "private": true,
  "scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev",
    "generate": "nuxt generate",
    "preview": "nuxt preview"
  },
  "devDependencies": {
    "@nuxt/types": "^2.15.8",
    "@nuxt/typescript-build": "^2.1.0",
    "nuxt": "3.0.0-rc.3"
  },
  "dependencies": {
    "@oruga-ui/oruga": "^0.5.5",
    "@oruga-ui/oruga-next": "^0.5.5",
    "apollo-cache-inmemory": "^1.6.6",
    "apollo-client": "^2.6.10",
    "apollo-link-http": "^1.5.17",
    "apollo-link-ws": "^1.0.20",
    "apollo-utilities": "^1.3.4",
    "babel-eslint": "^10.1.0",
    "eslint": "5",
    "eslint-plugin-vue": "^9.0.1",
    "graphql": "^16.5.0",
    "graphql-tag": "^2.12.6",
    "graphql-ws": "^5.8.2",
    "stylus": "^0.57.0",
    "subscriptions-transport-ws": "^0.11.0",
    "typescript-eslint": "^0.0.1-alpha.0",
    "underscore": "^1.13.3",
    "vue": "latest",
    "vuex": "^4.0.2"
  },
  "eslintConfig": {
    "root": true,
    "env": {
      "node": true
    },
    "extends": [
      "plugin:vue/essential",
      "eslint:recommended"
    ],
    "parserOptions": {
      "parser": "babel-eslint"
    },
    "rules": {}
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "not dead"
  ]
}
could you help me guys?

**Reply by pi0** [reply]
Without reproduction, it is almost impossible to guess. What modules are you registering in `nuxt.config`?

**Reply by richardeschloss** [reply]
Ok, I was able to reproduce this issue in a very odd way. But the problem / fix is embarrassingly simple. 
If someone has a nuxt config and they forget to use a comma (",") between the `modules` entries, this error will be encountered.

nuxt.config
```js
  modules: [
    '/mod1'  // <-- no comma, but there should be one...
    [ '/mod2' ]
  ]
```

So, if the config is improperly formed javascript, the error message thrown from the above config is "Cannot start nuxt:  Nuxt module should be a function: undefined" when really, it's improper javascript. This is fixable by the user... just insert a comma between entries.

**Comment by archx3** [comment]
Hi
I'm trying to use [https://tabler.io/docs/icons/vue](tabler icons vue) and I keep ketting 
` ERROR`  `Cannot restart nuxt:  Nuxt module should be a function: [object Object] `

This is how my nuxt modules looks like
```
{
  ...,
  modules: [
    '@nuxt/image-edge',
    '@tabler/icons-vue',
  ],
  ...
}
```

**Reply by danielroe** [reply]
You probably have a file in your `~/modules` directory which does not have a default export of a Nuxt module. (The directory is now auto-scanned - a feature in v3.3. You can change the directory by setting `dir.modules`.)

**Comment by JPustkuchen** [comment]
I also ran into the error:
> ERROR  Cannot start nuxt:  Nuxt module should be a function: [object Object]

with this backtrace:
```
  at normalizeModule (node_modules/@nuxt/kit/dist/index.mjs:478:11)
  at installModule (node_modules/@nuxt/kit/dist/index.mjs:450:47)
  at initNuxt (node_modules/nuxt/dist/index.mjs:2469:13)
  at async load (node_modules/nuxi/dist/chunks/dev.mjs:6843:9)
  at async Object.invoke (node_modules/nuxi/dist/chunks/dev.mjs:6873:5)
  at async _main (node_modules/nuxi/dist/cli.mjs:49:20)
```
*Sadly, neither the error message nor the backtrace is very helpful finding the root cause.*

I had to comment out modules one by one and finally found
`"@sidebase/nuxt-parse"` (https://github.com/sidebase/nuxt-parse)
caused the issue. It should have never been added as nuxt module, as it's just a package, but not a nuxt module. It would be perfect if the error or backtrace could show which module isn't a module as expected.

@danielroe is there a chance perhaps to make it easier to track down the reasons for this kind of issue by exposing more clearly where / why it happens? In my case, which module isn't providing the nuxt module as expected?

**Reply by danielroe** [reply]
Yes, we should do that. I would be grateful for an issue or PR to track 🙏

**Reply by JPustkuchen** [reply]
@danielroe thanks for the lightning fast reply :) I'll put it on my list and link this issue!
(Not yet experienced enough with nuxt to be able to create a MR, I guess - coming from PHP / Drupal world - but I'll do my best ;) Thank you for all this **great** work!)

**Reply by richardeschloss** [reply]
@JPustkuchen to put you on the right track, here's the line of code you'd probably want to look at for gracefully installing modules: 
https://github.com/nuxt/nuxt/blob/53d902ad5e8d29efdf7670bfbac1e1ff3dfba5cf/packages/kit/src/module/install.ts#L42

It may be a good opportunity to also update the error type from `unknown` to `InstallationError` (or similar)
https://github.com/nuxt/nuxt/blob/53d902ad5e8d29efdf7670bfbac1e1ff3dfba5cf/packages/kit/src/module/install.ts#L43 (consider defaulting to `unknown` only when the error is truly unknown)

**Comment by onlyabdiel** [comment]
I am having an issue migrating `nuxtjs/sitemap` to `nuxt-simple-sitemap`

When I run `npz nuxi prepare` it logs this error
```log
ERROR  Nuxt module should be a function: nuxt-simple-sitemap            6:02:21 PM

  at loadNuxtModuleInstance (node_modules/@nuxt/kit/dist/index.mjs:2498:11)
  at async installModule (node_modules/@nuxt/kit/dist/index.mjs:2437:47)
  at async initNuxt (node_modules/nuxt/dist/index.mjs:3822:7)
  at async loadNuxt (node_modules/nuxt/dist/index.mjs:3920:5)
  at async loadNuxt (node_modules/@nuxt/kit/dist/index.mjs:2600:19)
  at async Object.run (node_modules/nuxi/dist/chunks/prepare.mjs:56:18)
  at async runCommand$1 (node_modules/nuxi/dist/shared/nuxi.4fde776c.mjs:1648:16)
  at async runCommand$1 (node_modules/nuxi/dist/shared/nuxi.4fde776c.mjs:1639:11)
  at async runMain$1 (node_modules/nuxi/dist/shared/nuxi.4fde776c.mjs:1773:7) 
```

`nuxt.config.js`

```js
import { defineNuxtConfig } from 'nuxt/config'

export default defineNuxtConfig({
  // Target: https://go.nuxtjs.dev/config-target
  target: 'static',
  generate: {
    fallback: true
  },

  // Global page headers: https://go.nuxtjs.dev/config-head
  head: {
    title: 'example.com',
    htmlAttrs: {
      lang: 'en'
    },
    meta: [
      { charset: 'utf-8' },
      { name: 'viewport', content: 'width=device-width, initial-scale=1' },
      { hid: 'description', name: 'description', content: '' },
      { name: 'format-detection', content: 'telephone=no' }
    ],
    link: [
      { rel: 'icon', type: 'image/x-icon', href: '/favicon.ico' }
    ],
    script: [
      { src: 'https://kit.fontawesome.com/311fdf5135.js', crossorigin: 'anonymous' }
    ]
  },
  /*
  ** Customize the progress bar color
  */
  loading: { color: '#428db9' },

  // Global CSS: https://go.nuxtjs.dev/config-css
  css: [
    '~/assets/sass/main.scss'
  ],

  // Plugins to run before rendering page: https://go.nuxtjs.dev/config-plugins
  plugins: [
  ],

  // Auto import components: https://go.nuxtjs.dev/config-components
  components: true,

  // Modules for dev and build (recommended): https://go.nuxtjs.dev/config-modules
  buildModules: [
    // https://go.nuxtjs.dev/typescript
  ],

  // Modules: https://go.nuxtjs.dev/config-modules
  modules: [
    // https://go.nuxtjs.dev/content
    '@nuxt/content',
    '@nuxtjs/google-fonts',
    '@nuxtjs/i18n',
    'nuxt-simple-sitemap',
    '@nuxt/typescript-build'
  ],

  googleFonts: {
    display: 'swap',
    base64: true,
    families: {
      Rubik: true
    }
  },
  i18n: {
    baseUrl: 'https://example.com',
    locales: [
      {
        code: 'en',
        iso: 'en-US' // Will be used as catchall locale by default
      },
      {
        code: 'es',
        iso: 'es-MX'
      },
      {
        code: 'de',
        iso: 'de-DE'
      }
    ],
    parsePages: false,
    pages: {
      about: {
        es: '/quien-es-s-campos',
        en: '/who-is-s-campos',
        de: '/wer-ist-s-campos'
      }
    },
    defaultLocale: 'es',
    strategy: 'prefix_except_default',
    vueI18n: './nuxt-i18n.ts'
  },
  sitemap: {
    hostname: 'https://example.com',
    i18n: true,
    defaults: {
      changefreq: 'daily',
      priority: 1,
      lastmod: new Date()
    }
  },
  site: {
    url: 'https://example.com'
  },

  // Content module configuration: https://go.nuxtjs.dev/config-content
  content: {},

  // Build Configuration: https://go.nuxtjs.dev/config-build
  build: {
  }
})
```

`package.json`

```json
{
  "name": "example.com",
  "version": "1.0.0",
  "private": true,
  "scripts": {
    "dev": "nuxt dev",
    "build": "nuxt build",
    "start": "nuxt start",
    "generate": "nuxt generate",
    "lint:js": "eslint --ext \".js,.ts,.vue\" --ignore-path .gitignore .",
    "lint": "npm run lint:js",
    "lintfix": "npm run lint:js -- --fix",
    "test": "jest"
  },
  "dependencies": {
    "@nuxt/content": "^2.0.0",
    "@nuxtjs/i18n": "^8.0.0",
    "core-js": "^3.19.3",
    "nuxt": "^3.8.2",
    "vue": "^3.4.0",
    "vue-server-renderer": "^2.6.14",
    "vue-template-compiler": "^2.6.14",
    "webpack": "^4.46.0"
  },
  "devDependencies": {
    "@babel/eslint-parser": "^7.16.3",
    "@nuxt/types": "^2.15.8",
    "@nuxt/typescript-build": "^2.1.0",
    "@nuxtjs/eslint-config-typescript": "^8.0.0",
    "@nuxtjs/eslint-module": "^3.0.2",
    "@nuxtjs/google-fonts": "^3.0.0-0",
    "@vue/test-utils": "^1.3.0",
    "babel-core": "7.0.0-bridge.0",
    "babel-jest": "^27.4.4",
    "eslint": "^8.4.1",
    "eslint-plugin-nuxt": "^3.1.0",
    "eslint-plugin-vue": "^8.2.0",
    "jest": "^27.4.4",
    "nuxt-simple-sitemap": "^4.2.1",
    "sass": "^1.54.5",
    "sass-loader": "10.1.1",
    "ts-jest": "^27.1.1",
    "vue-jest": "^3.0.4"
  }
}
```

**Reply by damevin** [reply]
Still have this issue ?

**Reply by onlyabdiel** [reply]
I don't have the issue anymore. It was an issue with the packages. I tried commenting the modules elements array one by one. When I commented typescript module the issue disappeared.

Thanks for the answers!

**Reply by TheAlexLichter** [reply]
(The typescript module is not Nuxt 3 compatible)

**Comment by RGBvision** [comment]
> npm i
> postinstall
> nuxt prepare


 ERROR  Nuxt module should be a function: @element-plus/nuxt                                                                                                                                                                                                                       22:05:04

  at loadNuxtModuleInstance (node_modules/@nuxt/kit/dist/index.mjs:2498:11)
  at async installModule (node_modules/@nuxt/kit/dist/index.mjs:2437:47)
  at async initNuxt (node_modules/nuxt/dist/index.mjs:3822:7)
  at async loadNuxt (node_modules/nuxt/dist/index.mjs:3920:5)
  at async loadNuxt (node_modules/@nuxt/kit/dist/index.mjs:2600:19)
  at async Object.run (node_modules/nuxi/dist/chunks/prepare.mjs:56:18)
  at async runCommand$1 (node_modules/nuxi/dist/shared/nuxi.4fde776c.mjs:1648:16)
  at async runCommand$1 (node_modules/nuxi/dist/shared/nuxi.4fde776c.mjs:1639:11)
  at async runMain$1 (node_modules/nuxi/dist/shared/nuxi.4fde776c.mjs:1773:7)
  
```
"devDependencies": {
    "@element-plus/nuxt": "^1.0.7",
    "@nuxt/devtools": "latest",
    "@nuxtjs/color-mode": "^3.3.2",
    "@nuxtjs/i18n": "^8.0.0",
    "@types/bootstrap": "^5.2.10",
    "@types/nodemailer": "^6.4.14",
    "nuxt": "^3.9.0",
    "nuxt-gtag": "^1.1.2",
    "nuxt-og-image": "^3.0.0-rc.21",
    "nuxt-security": "^1.0.0",
    "nuxt-simple-sitemap": "^4.2.1",
    "sass": "^1.69.5",
    "sass-loader": "^13.3.2",
    "typescript": "^5.3.3",
    "vue": "^3.4.0",
    "vue-router": "^4.2.5"
  },
  "dependencies": {
    "@formkit/auto-animate": "^0.8.1",
    "@motionone/vue": "^10.16.4",
    "@nuxt/content": "^2.10.0",
    "@nuxt/image": "^1.1.0",
    "@popperjs/core": "^2.11.8",
    "@vueuse/core": "^10.7.0",
    "@vueuse/motion": "^2.0.0",
    "bootstrap": "^5.3.2",
    "bowser": "^2.11.0",
    "element-plus": "^2.4.4",
    "libphonenumber-js": "^1.10.52",
    "nodemailer": "^6.9.7",
    "vue-easy-lightbox": "^1.17.0",
    "vue3-carousel-nuxt": "^1.1.0"
  }
```

The problem appears in Ubuntu 20.04. There are no problems on Windows 11. Both machines have Node v21.5.0 installed.

**Reply by TheAlexLichter** [reply]
Elements plus needs a major bump, v1 does not support Nuxt 3.

**Comment by Rednas83** [comment]
When starting the dev server with nuxt 3.14.0 I get
`ERROR  Cannot start nuxt:  Nuxt module should be a function: @nuxtjs/color-mode `

```
------------------------------
- Operating System: Windows_NT
- Node Version:     v22.11.0
- Nuxt Version:     3.14.0
- CLI Version:      3.15.0
- Nitro Version:    2.10.2
- Package Manager:  pnpm@9.12.3
- Builder:          -
- User Config:      default
- Runtime Modules:  normalizedModule(), normalizedModule(), @vueuse/nuxt@11.2.0, @nuxt/content@2.13.4, @nuxt/eslint@0.6.1, @nuxtjs/tailwindcss@6.12.2, @nuxt/devtools@1.6.0, shadcn-nuxt@0.11.2, nuxt-meilisearch@1.3.10, @nuxt/image@1.8.1, @formkit/nuxt@1.6.7, @nuxt/icon@1.6.1
- Build Modules:    -
------------------------------
```

**Comment by tc0o** [comment]
If you're encountering the same issue with Docker Compose and the `NODE_ENV=production` environment variable, keep in mind that some modules may not be installed. You may need to install all the necessary dependencies. To do this, run:

```bash
npm install --include=dev
```
